File: screens/habits.py
"""The habit strip on the dashboard — one tap ticks a habit off for today.

Renders a block into the deck browser and hands `web/habits/habits.js` the data
it needs to redraw a chip on its own. Clicks never re-render the page: a full
`deckBrowser.refresh()` per tick would drop the scroll position and flash the
whole screen for a checkbox.

The chip's next value is worked out twice on purpose — optimistically in JS so
the tap feels instant, then authoritatively by `HabitStore.toggle`, whose answer
is pushed back through `AwdHabit.applyValue`. The store is the truth; the JS copy
only has to be right often enough that the correction is invisible.
"""

import html
import json

from aqt import mw

from ..core.translations import tr
from ..features.habits import scheduling, stats
from ..features.habits.store import get_store

# How far back a streak may reach on the dashboard. Two years of chips is
# already more history than the strip can say anything about, and it bounds
# the work done on every render.
STREAK_WINDOW = 730


def _streak_for(store, habit, today: int, first_day: int) -> dict:
    values = store.values(habit.id, scheduling.shift(today, -STREAK_WINDOW), today)
    return stats.streaks(habit, values, today, first_day)


def _week_done(store, habit, today: int, first_day: int) -> int:
    start = scheduling.week_start(today, first_day)
    return stats.week_count(
        habit, store.values(habit.id, start, scheduling.shift(start, 6)), start
    )


def habit_payload(store, habits: list, today: int, first_day: int) -> list:
    """Everything the strip needs about each habit, as plain JSON values."""
    day_values = store.get_day(today)
    payload = []
    for habit in habits:
        value = int(day_values.get(habit.id, 0) or 0)
        streak = _streak_for(store, habit, today, first_day)
        entry = {
            "id": habit.id,
            "name": habit.name,
            "icon": habit.icon,
            "color": habit.color,
            "count": habit.is_count,
            "target": habit.target,
            "unit": habit.unit,
            "step": habit.step(),
            "value": value,
            "due": scheduling.is_due(habit, today),
            "streak": streak["current"],
            "weekly": streak["weekly"],
        }
        if streak["weekly"]:
            entry["weekDone"] = _week_done(store, habit, today, first_day)
            entry["weekTarget"] = scheduling.weekly_target(habit)
        payload.append(entry)
    return payload


def data_for_web() -> dict:
    """`AWD_DATA.habits`, or an empty payload when the block is switched off."""
    store = get_store()
    today = store.today()
    first_day = store.first_weekday()
    habits = store.habits()
    week_done = {
        habit.id: _week_done(store, habit, today, first_day)
        for habit in habits
        if scheduling.is_weekly(habit)
    }
    summary = stats.day_summary(habits, store.get_day(today), today, week_done)
    return {
        "today": today,
        "items": habit_payload(store, habits, today, first_day),
        "done": summary["done"],
        "due": summary["due"],
        "i18n": {
            "done": tr("habit_done_of"),
            "allDone": tr("habit_all_done"),
            "offDay": tr("habit_off_day"),
            "days": tr("days_unit"),
            "weeks": tr("weeks_unit"),
            "thisWeek": tr("habit_week_progress"),
        },
    }


def card_html() -> str:
    """The block itself. The chips are filled in by habits.js from AWD_DATA."""
    from .dashboard import icon

    actions = (
        f'<button class="awd-pill awd-pill-ghost awd-hb-action"'
        f' onclick="pycmd(\'awd:habit:report\')">'
        f'{icon("chart")}<span>{html.escape(tr("habit_report"))}</span></button>'
        f'<button class="awd-pill awd-pill-ghost awd-pill-icon awd-hb-action"'
        f' title="{html.escape(tr("habit_manage"))}"'
        f' onclick="pycmd(\'awd:habit:manage\')">{icon("sliders")}</button>'
    )
    return f"""
    <section class="awd-card awd-hb-card">
      <div class="awd-card-head">
        <span class="awd-chip">{html.escape(tr("habits"))}</span>
        <div class="awd-hb-head">
          <span class="awd-hb-count" id="awd-hb-count"></span>
          {actions}
        </div>
      </div>
      <div class="awd-hb-bar"><i id="awd-hb-fill"></i></div>
      <div class="awd-hb-grid" id="awd-hb-grid"></div>
      <div class="awd-hb-empty" id="awd-hb-empty" hidden>
        <div class="awd-empty-title">{html.escape(tr("habit_empty"))}</div>
        <div class="awd-empty-hint">{html.escape(tr("habit_empty_hint"))}</div>
        <button class="awd-pill awd-pill-accent"
                onclick="pycmd('awd:habit:manage')">
          {icon("plus")}<span>{html.escape(tr("habit_add"))}</span></button>
      </div>
    </section>
    """


# --- bridge commands -----------------------------------------------------------

def _push(habit_id: str, value: int) -> None:
    """Send the stored value back to the page so it can correct itself."""
    try:
        payload = json.dumps({"id": habit_id, "value": value})
        mw.deckBrowser.web.eval(f"window.AwdHabit && AwdHabit.applyValue({payload});")
    except Exception as e:
        print(f"[Awesome Dashboard] habits: cannot push value back: {e}")


def toggle(habit_id: str) -> None:
    store = get_store()
    _push(habit_id, store.toggle(habit_id, store.today()))


def open_manager() -> None:
    from ..ui.habits import open_manager as show

    show(mw)


def open_report() -> None:
    from . import habit_report

    habit_report.open_overlay()

File: web/habits/habits.js
/* ============================================================
   habits.js — the habit strip on the dashboard.
   Reads window.AWD_DATA.habits injected by screens/habits.py.

   A tap repaints its own chip and nothing else. Re-rendering the
   deck browser for a checkbox would drop the scroll position and
   flash the page, so the DOM is updated optimistically and the
   value the store actually wrote comes back through applyValue().
   ============================================================ */

(function () {
  "use strict";

  var AwdHabit = (window.AwdHabit = window.AwdHabit || {});
  var items = [];
  var byId = {};

  function data() {
    return (window.AWD_DATA && window.AWD_DATA.habits) || {};
  }

  function i18n(key) {
    return (data().i18n || {})[key] || "";
  }

  function fill(text, values) {
    return String(text).replace(/\{(\w+)\}/g, function (whole, name) {
      return name in values ? values[name] : whole;
    });
  }

  function escapeHtml(text) {
    return String(text).replace(/[&<>"']/g, function (c) {
      return { "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" }[c];
    });
  }

  /* ---------- model ----------
     `base` is the streak *without* today counted, so the badge can move the
     moment the chip does. Appending today to a run only ever adds one, and a
     weekly habit's run only moves when the week crosses its target — anything
     subtler than that is the next full render's job. */

  function isComplete(item) {
    return item.count ? item.value >= item.target : item.value > 0;
  }

  function weekComplete(item) {
    return !!item.weekly && item.weekDone >= item.weekTarget;
  }

  /* What the header counts. A weekly habit that has already made its week is
     not an outstanding task today — see stats.day_summary, which counts the
     same way on the Python side. */
  function satisfied(item) {
    return isComplete(item) || weekComplete(item);
  }

  function prepare(list) {
    items = (list || []).map(function (raw) {
      var item = {};
      for (var key in raw) item[key] = raw[key];
      var counted = item.weekly ? weekComplete(item) : item.due && isComplete(item);
      item.base = Math.max(0, item.streak - (counted ? 1 : 0));
      return item;
    });
    byId = {};
    items.forEach(function (item) {
      byId[item.id] = item;
    });
  }

  function streakOf(item) {
    var counted = item.weekly ? weekComplete(item) : item.due && isComplete(item);
    return item.base + (counted ? 1 : 0);
  }

  /* Mirror of HabitStore.toggle. The store stays the authority — this only has
     to agree with it often enough that applyValue() never visibly corrects. */
  function nextValue(item) {
    if (!item.count) return item.value ? 0 : 1;
    if (item.value >= item.target) return 0;
    return Math.min(item.target, item.value + (item.step || 1));
  }

  /* ---------- rendering ---------- */

  function subText(item) {
    if (!item.due) return i18n("offDay");
    if (item.count) {
      return item.value + "/" + item.target + (item.unit ? " " + item.unit : "");
    }
    if (item.weekly) {
      return fill(i18n("thisWeek"), { done: item.weekDone, total: item.weekTarget });
    }
    return "";
  }

  function chipHtml(item) {
    var sub = subText(item);
    // No decrement control. It was absolutely positioned over the tick circle,
    // so hovering a count habit looked like the tick had turned into a minus.
    // Tapping past the target already wraps back to zero.
    return (
      '<span class="awd-hb-ic">' + escapeHtml(item.icon) + "</span>" +
      '<span class="awd-hb-text">' +
      '<span class="awd-hb-name">' + escapeHtml(item.name) + "</span>" +
      '<span class="awd-hb-sub">' + escapeHtml(sub) + "</span>" +
      "</span>" +
      '<span class="awd-hb-streak"></span>' +
      '<span class="awd-hb-check">✓</span>'
    );
  }

  function paintChip(node, item) {
    var complete = isComplete(item);
    var partial = !complete && item.value > 0;
    node.classList.toggle("done", complete);
    node.classList.toggle("partial", partial);
    node.classList.toggle("satisfied", !complete && weekComplete(item));
    node.classList.toggle("off", !item.due);
    node.style.setProperty(
      "--awd-fill",
      item.count && item.target ? (100 * Math.min(1, item.value / item.target)).toFixed(1) + "%" : "0%"
    );
    var sub = node.querySelector(".awd-hb-sub");
    if (sub) sub.textContent = subText(item);
    // Just the number: the unit costs more width than a chip can spare, and
    // spending it here is what pushes the habit's own name into an ellipsis.
    var run = streakOf(item);
    var unit = item.weekly ? i18n("weeks") : i18n("days");
    var streak = node.querySelector(".awd-hb-streak");
    if (streak) streak.textContent = run ? "🔥 " + run : "";

    var title = item.name;
    if (run) title += " · 🔥 " + run + " " + unit;
    if (item.weekly) {
      title += " · " + fill(i18n("thisWeek"), {
        done: item.weekDone,
        total: item.weekTarget,
      });
    }
    node.title = title;
  }

  function paintHeader() {
    var due = 0;
    var done = 0;
    items.forEach(function (item) {
      if (!item.due) return;
      due += 1;
      if (satisfied(item)) done += 1;
    });
    var label = document.getElementById("awd-hb-count");
    if (label) {
      label.classList.toggle("all-done", due > 0 && done === due);
      // Escape the template first, then substitute — the placeholders take
      // numbers and the <b> is ours, so nothing user-supplied reaches innerHTML.
      label.innerHTML =
        due === 0
          ? ""
          : done === due
            ? escapeHtml(i18n("allDone"))
            : fill(escapeHtml(i18n("done")), {
                done: "<b>" + done + "</b>",
                total: due,
              });
    }
    var bar = document.getElementById("awd-hb-fill");
    if (bar) bar.style.width = due ? (100 * done / due).toFixed(1) + "%" : "0%";
  }

  function render() {
    var host = document.getElementById("awd-hb-grid");
    if (!host) return;
    host.innerHTML = "";
    items.forEach(function (item) {
      var chip = document.createElement("button");
      chip.className = "awd-hb";
      chip.dataset.id = item.id;
      chip.style.setProperty("--awd-tint", item.color);
      chip.innerHTML = chipHtml(item);
      chip.addEventListener("click", function (event) {
        AwdHabit.tick(event, item.id);
      });
      host.appendChild(chip);
      paintChip(chip, item);
    });
    var card = host.closest(".awd-hb-card");
    var empty = document.getElementById("awd-hb-empty");
    if (empty) empty.hidden = items.length > 0;
    if (card) card.dataset.empty = items.length ? "0" : "1";
    paintHeader();
  }

  function nodeFor(id) {
    return document.querySelector('.awd-hb[data-id="' + id + '"]');
  }

  function commit(item, value) {
    var was = isComplete(item);
    item.value = value;
    var now = isComplete(item);
    // Only a change of state moves the week's tally: stepping a count habit
    // from 1 to 2 out of 3 is not another repetition.
    if (item.weekly && now !== was) {
      item.weekDone = Math.max(0, (item.weekDone || 0) + (now ? 1 : -1));
    }
    var node = nodeFor(item.id);
    if (node) paintChip(node, item);
    paintHeader();
  }

  /* ---------- commands ---------- */

  AwdHabit.tick = function (event, id) {
    if (event) event.stopPropagation();
    var item = byId[id];
    if (!item) return;
    commit(item, nextValue(item));
    if (typeof pycmd === "function") pycmd("awd:habit:toggle:" + id);
  };

  /* What the store actually wrote. Normally identical to the optimistic value;
     when it is not, this is the correction. */
  AwdHabit.applyValue = function (payload) {
    var item = payload && byId[payload.id];
    if (!item || item.value === payload.value) return;
    commit(item, payload.value);
  };

  /* ---------- init ---------- */

  function init() {
    if (!document.getElementById("awd-hb-grid")) return;
    prepare(data().items);
    render();
  }

  if (document.readyState === "loading") {
    document.addEventListener("DOMContentLoaded", init);
  } else {
    init();
  }
})();
